jkube-kit/config/resource/src/main/java/org/eclipse/jkube/kit/config/resource/ResourceConfig.java

```diff
@@ -76,13 +108,74 @@ public class ResourceConfig {
   private Boolean createExternalUrls;
   private IngressConfig ingress;
   private String routeDomain;
+  /**
+   * @deprecated Use nested controller configuration instead
+   */
+  @Deprecated
   private String restartPolicy;
+  private ControllerResourceConfig controller;
+  private List<ControllerResourceConfig> controllers;
 
   public static ResourceConfigBuilder toBuilder(ResourceConfig original) {
     return Optional.ofNullable(original).orElse(new ResourceConfig()).toBuilder();
   }
 
-    // TODO: SCC
+  public ControllerResourceConfig getController() {
+    if (controller == null) {
+      controller = createNewControllerConfig();
+    }
+    return controller;
+  }
+
+  public List<ControllerResourceConfig> getControllers() {
+    if (controllers == null || controllers.isEmpty()) {
+      controllers = new ArrayList<>();
+      controllers.add(createNewControllerConfig());
+    }
+    return controllers;
+  }
```

Comment:
_Still reviewing the rest of the code_

Not sure how we're dealing with the case where a user specifies both controller and controllers. Or the case where the user is specifying all of the other options.

IMHO `getController` should remain private and access to controller information should only be allowed through the list (I think we should be doing something similar elsewhere).

Following are the cases I can gather:

1. No controller, no controllers, no top-level controller settings:
   We might want to init a dummy Controller object to avoid NPE, but this is not very elegant (`getControllers` returns a single empty element)
2. No controller, no controllers, but top-level controller settings added:
   `getControllers` returns the result of `createNewControllerConfig`
3. Controller, no controllers, no top-level controller settings:
   `getControllers` returns a single element with the provided controller
4. No controller, controllers, no top-level controller settings
   `getControllers` returns the list of controllers
5. Controller, controllers, no top-level controller settings
   `getControllers` returns a merged list of controllers prepending the single controller
6. Controller or controllers, and top-level controller settings
   `getControllers` returns a merged list of controllers prepending the single.
   top-level settings are ignored since they are deprecated (we might want to warn the user)

Reply:
After discussion
Remove the `getControllers` method and just keep a single `getController` that's in charge of retrieving or merging the legacy settings. i.e. we don't support multiple controllers.

If both (controller field and legacy fields are present), throw an exception with a helpful message for users to be able to safely migrate to the new config.
In case user is using legacy fields we log a debug message.